src/SubspaceClustering.jl

```diff
@@ -2,4 +2,85 @@ module SubspaceClustering
 
 # Write your package code here.
 
+## K-Subspaces SubspaceClustering
+
+#Imports
+import LinearAlgebra: norm, svd, transpose
+import Base: copy, deepcopy, findall, view
+import Base:argmax
+using StableRNGs: randn
+using Logging: @info, @warn
+using ProgressLogging: @progress
+using ArnoldiMethod: partialschur
+
+#Exports
+export polar, KSS
+
+#Polar function
+
+function polar(X)
+	U, _, V = svd(X)
+	U*V'
+end
+
+"""
+	KSS(X, d; niters=100, Uinit=nothing)
+
+Run K-subspaces on the data matrix `X`
+with subspace dimensions `d[1], ..., d[K]`.
+
+# Arguments
+- `X::AbstractMatrix`: Data matrix of size `(D, N)`. Each column represents a data point.
+- `d::Vector{Int}`: A vector of length `K` containing the subspace dimensions for each of the 'K' subspaces.
+
+# Keyword Arguments
+- `niters::Int=100`: Maximum number of iterations (default is 100).
+-'Uinit::Union{Nothing, Vector{AbstractMatrix}}': Initial subspaces for the algorithm. If set to nothing, random subspaces are initialized.
+
+# Returns
+- 'U::Vector{AbstractMatrix}': A vector of length `K` containing the subspace basis matrices for 'K' clusters.
+- 'c::Vector{Int}': A vector of length `N` containing the cluster assignments for each data point.
+"""
+function KSS(X, d; niters=100, Uinit=polar.(randn.(size(X, 1), collect(d))))
+	K = length(d)
+	D, N = size(X)
+
+	# Initialize
+	U = deepcopy(Uinit)
+	c = [argmax(norm(U[k]' * view(X, :, i)) for k in 1:K) for i in 1:N]
+	c_prev = copy(c)
+
+	# Iterations
+	@progress for t in 1:niters
+		# Update subspaces
+		for k in 1:K
+			ilist = findall(==(k), c)
+			# println("Cluster $k, ilist size: ", length(ilist))
+			if isempty(ilist)
+				@warn "Cluster $k is empty; re-initializing subspace."
```

Comment:
Don't think this needs to be a `@warn`. Can you check what is done in `Clustering.jl`?

Reply:
This is from Clustering.jl package, "test/clustering_quality.jl". They are using test_logs macro to generate the message. There is also a test_warn macro. 



```
    @testset "empty clusters" begin
        # degenerated clustering, no 4th cluster
        degenC = [0 2  0 -2 -2
                  4 0 -4  0  0]
        degenA = [1, 1, 2, 2, 2, 2, 3, 3, 5, 5, 5, 5]

        @test_logs((:warn, "Detected empty cluster(s): #4. clustering_quality() results might be incorrect."),
                   clustering_quality(Y, degenC, degenA; quality_index = :calinski_harabasz))
        @test clustering_quality(Y, degenC, degenA; quality_index = :calinski_harabasz, metric = Euclidean()) ≈ (32/3) / (16/8)
    end
```

Reply:
Got it. Don't want to spam the user with messages when clusters get re-initialized but I think it's good to alert them. Maybe can be handled by suggesting that they consider reducing the number of clusters. Let's change the warning as follows:

```suggestion
				@warn "Empty cluster detected at iteration $t - reinitializing the subspace. Consider reducing the number of clusters."
```

BTW, `@test_logs` and `@test_warn` are not used to create the logs but to test them. See here: https://docs.julialang.org/en/v1/stdlib/Test/